maas/resource_maas_zone.go

```diff
@@ -0,0 +1,142 @@
+package maas
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/canonical/gomaasclient/client"
+	"github.com/canonical/gomaasclient/entity"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceMaasZone() *schema.Resource {
+	return &schema.Resource{
+		Description:   "Provides a resource to manage MAAS zones.",
+		CreateContext: resourceZoneCreate,
+		ReadContext:   resourceZoneRead,
+		UpdateContext: resourceZoneUpdate,
+		DeleteContext: resourceZoneDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: func(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+				client := meta.(*client.Client)
+				zone, err := getZone(client, d.Id())
+				if err != nil {
+					return nil, err
+				}
+				tfState := map[string]interface{}{
+					"name":        zone.Name,
+					"description": zone.Description,
+				}
+				if err := setTerraformState(d, tfState); err != nil {
+					return nil, err
+				}
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+
+		Schema: map[string]*schema.Schema{
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "A brief description of the new zone.",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The name of the new zone.",
+			},
+		},
+	}
+}
+
+func resourceZoneRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
+	zone, err := client.Zone.Get(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	tfstate := map[string]any{
+		"name":        zone.Name,
+		"description": zone.Description,
+	}
+
+	if err := setTerraformState(d, tfstate); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceZoneCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
+	params := getZoneParams(d)
+	zone, err := findZone(client, params.Name)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if zone == nil {
+		zone, err = client.Zones.Create(params)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+	d.SetId(zone.Name)
+
+	return resourceZoneUpdate(ctx, d, meta)
+}
+
+func resourceZoneUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client.Client)
+
```

Comment:
This is needed since we want to save the database ID. Then `id` should be given to `Zone.Update` instead of `d.Id()`.

```suggestion
id, err := strconv.Atoi(d.Id())
	if err != nil {
		return diag.FromErr(err)
	}
```